planning/planning_validator/src/planning_validator.cpp

```diff
@@ -0,0 +1,451 @@
+  s.is_valid_longitudinal_min_acc = checkValidMinLongitudinalAcceleration(trajectory);
+  s.is_valid_velocity_deviation = checkValidVelocityDeviation(trajectory);
+  s.is_valid_distance_deviation = checkValidDistanceDeviation(trajectory);
+
+  // use resampled trajectory because the following metrics can not be evaluated for closed points.
+  // Note: do not interpolate to keep original trajectory shape.
+  constexpr auto min_interval = 1.0;
+  const auto resampled = resampleTrajectory(trajectory, min_interval);
+
+  s.is_valid_relative_angle = checkValidRelativeAngle(resampled);
+  s.is_valid_curvature = checkValidCurvature(resampled);
+  s.is_valid_steering = checkValidSteering(resampled);
+  s.is_valid_steering_rate = checkValidSteeringRate(resampled);
+
+  s.invalid_count = isAllValid(s) ? 0 : s.invalid_count + 1;
+}
+
+bool PlanningValidator::checkValidFiniteValue(const Trajectory & trajectory)
+{
+  for (const auto & p : trajectory.points) {
+    if (!checkFinite(p)) return false;
+  }
+  return true;
+}
+
+bool PlanningValidator::checkValidInterval(const Trajectory & trajectory)
+{
+  const auto [max_interval_distance, i] = calcMaxIntervalDistance(trajectory);
+  validation_status_.max_interval_distance = max_interval_distance;
+
+  if (max_interval_distance > validation_params_.interval_threshold) {
+    if (i > 0) {
+      const auto & p = trajectory.points;
+      debug_pose_publisher_->pushPoseMarker(p.at(i - 1), "trajectory_interval");
+      debug_pose_publisher_->pushPoseMarker(p.at(i), "trajectory_interval");
+    }
+    return false;
+  }
+
+  return true;
+}
+
+bool PlanningValidator::checkValidRelativeAngle(const Trajectory & trajectory)
+{
+  const auto [max_relative_angle, i] = calcMaxRelativeAngles(trajectory);
+  validation_status_.max_relative_angle = max_relative_angle;
+
+  if (max_relative_angle > validation_params_.relative_angle_threshold) {
+    const auto & p = trajectory.points;
+    if (i < p.size() - 3) {
+      debug_pose_publisher_->pushPoseMarker(p.at(i), "trajectory_relative_angle", 0);
+      debug_pose_publisher_->pushPoseMarker(p.at(i + 1), "trajectory_relative_angle", 1);
+      debug_pose_publisher_->pushPoseMarker(p.at(i + 2), "trajectory_relative_angle", 2);
+    }
+    return false;
+  }
+  return true;
+}
+
+bool PlanningValidator::checkValidCurvature(const Trajectory & trajectory)
+{
+  const auto [max_curvature, i] = calcMaxCurvature(trajectory);
+  validation_status_.max_curvature = max_curvature;
+  if (max_curvature > validation_params_.curvature_threshold) {
+    const auto & p = trajectory.points;
+    if (i > 0 && i < p.size() - 1) {
+      debug_pose_publisher_->pushPoseMarker(p.at(i - 1), "trajectory_curvature");
+      debug_pose_publisher_->pushPoseMarker(p.at(i), "trajectory_curvature");
+      debug_pose_publisher_->pushPoseMarker(p.at(i + 1), "trajectory_curvature");
+    }
+    return false;
+  }
+  return true;
```

Comment:
(I feel you do not need to modify in this PR)

Just comment. I prefer
```cpp
boost::optional<size_t> calcMaxCurvatureIndex(Trajectory);
```

```cpp
  const auto idx = calcMaxCurvatureIndex(trajectory);
  if (!idx) {
    return true;
  }

  validation_status_.max_curvature = idx.get();
  if (max_curvature > validation_params_.curvature_threshold) {
    const auto & p = trajectory.points;
    debug_pose_publisher_->pushPoseMarker(p.at(i - 1), "trajectory_curvature");
    debug_pose_publisher_->pushPoseMarker(p.at(i), "trajectory_curvature");
    debug_pose_publisher_->pushPoseMarker(p.at(i + 1), "trajectory_curvature");
    return false;
  }
  return true;
```

since I disagree with returning `0` as not-enough points as follows.
```cpp
std::pair<double, size_t> calcMaxCurvature(const Trajectory & trajectory)
{
  if (trajectory.points.size() < 3) {
    return {0.0, 0};
  }
```

Reply:
The 0 curvature for trajectory sounds ok to me, it is based on the definition.
But yes, as you say, returning 0 index for the empty trajectory doesn't make sense.

Reply:
@takayuki5168 Let me think later, maybe updated in another PR. Please keep this comment open until that time.